From a9f66f6ddfc63fe6048ead62f21d9ac8aab2e5d4 Mon Sep 17 00:00:00 2001
From: Alexander Brauckmann <a.brauckmann@gmail.com>
Date: Tue, 21 Feb 2023 18:46:00 +0000
Subject: [PATCH] Disable TOSA to make mlir-hlo build

---
 CMakeLists.txt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index ae5e4e2f..a79fd510 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -172,7 +172,7 @@ add_subdirectory(stablehlo)
 add_subdirectory(tests)
 add_subdirectory(thlo)
 add_subdirectory(tools)
-add_subdirectory(tosa)
+#add_subdirectory(tosa)
 add_subdirectory(transforms)
 add_subdirectory(utils)
 
-- 
2.34.1

